File: python/largest_rectangle_in_histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        res = 0

        for i, h in enumerate(heights):
            begin = i
            while stack and stack[-1][0] > h:
                res = max(res, (i - stack[-1][1]) * stack[-1][0])
                begin = stack[-1][1]
                stack.pop()
            stack.append((h, begin))
        
        print(stack)
        while stack:
            res = max(res, (len(heights) - stack[-1][1])*stack[-1][0])
            stack.pop()
        return res

File: C++/largest_rectangle_in_histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int res = 0;
        stack<pair<int, int>> s;

        for (int i=0; i<heights.size(); i++){
            int begin = i;
            while (!s.empty() && s.top().first > heights[i]){
                res = max(res, (i-s.top().second) *  s.top().first);
                begin = s.top().second;
                s.pop();
            }

            pair<int, int> p = {heights[i], begin};
            s.push(p);
        }

        while (!s.empty()){
            res = max(res, ((int) heights.size()-s.top().second) * s.top().first);
            s.pop();
        }
        return res;
    }
};
